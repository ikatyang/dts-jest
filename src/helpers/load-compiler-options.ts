import * as fs from 'fs';
import * as path from 'path';
import * as ts from 'typescript';
import {ITSConfig} from '../definitions';

export const load_compiler_options = (tsconfig: ITSConfig, cwd: string = process.cwd()): ts.CompilerOptions => {
  const {
    compilerOptions: compiler_options = {},
  } = tsconfig;

  if (typeof tsconfig.extends === 'string') {
    const filename = path.resolve(cwd, tsconfig.extends);
    const {
      extends: target_extends,
      compilerOptions: parent_compiler_options = {},
    } = JSON.parse(fs.readFileSync(filename, 'utf-8'));

    return load_compiler_options(
      {
        extends: target_extends,
        compilerOptions: {
          ...parent_compiler_options,
          ...compiler_options,
        },
      },
      path.dirname(filename),
    );
  }

  return compiler_options;
};
