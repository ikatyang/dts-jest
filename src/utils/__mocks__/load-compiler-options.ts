export const load_compiler_options = () => ({});
