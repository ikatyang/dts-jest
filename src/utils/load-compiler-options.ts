import * as fs from 'fs';
import * as path from 'path';
import * as ts from 'typescript';
import {ITSConfig} from '../definitions';

export const load_compiler_options = (tsconfig: ITSConfig, cwd: string = ''): ts.CompilerOptions => {
  if (typeof tsconfig.extends === 'string') {
    const filename = path.resolve(cwd, tsconfig.extends);
    const {
      extends: target_extends,
      compilerOptions: target_compiler_options = {},
    } = JSON.parse(fs.readFileSync(filename, 'utf-8'));
    return load_compiler_options(
      {
        extends: target_extends,
        compilerOptions: {
          ...target_compiler_options,
          ...tsconfig.compilerOptions,
        },
      },
      path.dirname(filename),
    );
  }
  // tslint:disable-next-line:strict-boolean-expressions
  return tsconfig.compilerOptions || {};
};
