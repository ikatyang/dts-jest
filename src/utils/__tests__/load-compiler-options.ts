jest.unmock('../load-compiler-options.ts');

import * as fs from 'fs';
import * as path from 'path';
import {load_compiler_options} from '../load-compiler-options';

it('should return empty object while there is no key "extends" or "compilerOptions" in tsconfig', () => {
  const tsconfig = {some_option: 'some_value'};
  expect(load_compiler_options(tsconfig)).toEqual({});
});

it('should return property "compilerOptions" while there is no key "extends" but "compilerOptions" in tsconfig', () => {
  const tsconfig = {compilerOptions: {some_option: 'some_value'}};
  expect(load_compiler_options(tsconfig)).toBe(tsconfig.compilerOptions);
});

const load_compiler_options_in_file = (tsconfig_filename: string) => {
  const filename = path.resolve(__dirname, tsconfig_filename);
  const dirname = path.dirname(filename);
  const tsconfig = JSON.parse(fs.readFileSync(filename, 'utf8'));
  return load_compiler_options(tsconfig, dirname);
};

it('should return extended "compilerOptions" while there is key "extends" in tsconfig (empty-object)', () => {
  expect(
    load_compiler_options_in_file('../../../fixtures/load-compier-options/empty-object/tsconfig.json'),
  ).toMatchSnapshot();
});

it('should return extended "compilerOptions" while there is key "extends" in tsconfig (same-folder)', () => {
  expect(
    load_compiler_options_in_file('../../../fixtures/load-compier-options/same-folder/tsconfig.json'),
  ).toMatchSnapshot();
});

it('should return extended "compilerOptions" while there is key "extends" in tsconfig (different-folder)', () => {
  expect(
    load_compiler_options_in_file('../../../fixtures/load-compier-options/different-folder/tsconfig.json'),
  ).toMatchSnapshot();
});
