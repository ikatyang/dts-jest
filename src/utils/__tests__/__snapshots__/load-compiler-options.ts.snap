// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should return extended "compilerOptions" while there is key "extends" in tsconfig (different-folder) 1`] = `
Object {
  "tsconfig": "0",
  "tsconfig-0": "0",
  "tsconfig-1": "1",
  "tsconfig-2": "2",
}
`;

exports[`should return extended "compilerOptions" while there is key "extends" in tsconfig (empty-object) 1`] = `
Object {
  "tsconfig": "0",
}
`;

exports[`should return extended "compilerOptions" while there is key "extends" in tsconfig (same-folder) 1`] = `
Object {
  "tsconfig": "0",
  "tsconfig-0": "0",
  "tsconfig-1": "1",
  "tsconfig-2": "2",
}
`;
